### ISSUE TITLE
AKS/ACR integration stops picking up new images

### ISSUE THREAD

**Issue by UnwashedMeme** [weaveworks/flux#1808]
Bug Report:

* Flux version: 1.10.1
* helm-operator: 0.6.0
* `registry.acr.enabled: True`
* problem: new images in registry not discovered by flux

We're seeing some odd behavior right now where Flux stops finding new images in ACR repos based on the AKS credentials provided through `registry.acr.enabled`.  After the 1.10.1 release we enabled that integration and started removing `imagePullSecrets` from our HelmReleases. Everything appears to work fine but after a couple days Flux stops seeing new images.   I don't see anything apparently related in the logs.

If I restart the flux pod it all works correctly and we get a spate of `Auto-release ...`.

I'm suspicious that the login to ACR is expiring after some period of time and hence it fails to proceed.  Restarting the pod causes a new token to be generated and it then works for several days.

**Comment by UnwashedMeme** [comment]
Original work was in https://github.com/weaveworks/flux/pull/1694

**Comment by alanjcastonguay** [comment]
The content of /etc/kubernetes/azure.json does get rewritten with some periodicity, and the token can change. `GetAzureCloudConfigAADToken` intentionally re-reads the file every time it's called. @squaremo I believed that `registry.credsFor()` got called every time credentials are needed - does it get called less frequently (eg only once per container registry/repo/etc)?

**Comment by squaremo** [comment]
> I believed that `registry.credsFor()` got called every time credentials are needed - does it get called less frequently (eg only once per container registry/repo/etc)?

That's correct -- it'll ask for fresh credentials every time it refreshes a particular image repo; and, it requests the set of tags every time around (every few minutes), so we should expect to see authentication failures in the logs regularly, if the token is out of date.

> I don't see anything apparently related in the logs.

@UnwashedMeme Do you get messages that look like 

```
ts=2019-03-11T11:24:09.387590938Z caller=warming.go:364 component=warmer updated=quay.io/cortexproject/table-manager successful=1 attempted=1
```

or do those stop appearing?

You can trace the registry requests by setting the argument `--registry-trace` in fluxd (`--set registry.trace=true` in the Helm chart). That might tell us what's happening when fluxd attempts to get the set of tags for an image repo. Beware: it's a lot more log messages.

**Comment by alanjcastonguay** [comment]
@squaremo please clarify -- is the first half or the second half of that quoted sentence correct? It sure sounds like you're saying the first half is correct, but I want to make sure we're on the same page.

**Comment by alanjcastonguay** [comment]
@UnwashedMeme any progress on reproducing with `registry.trace=true`?

**Comment by squaremo** [comment]
`credsFor` is called each time an image repository (e.g., `quay.io/weave/works/flux`) is scanned.

**Comment by UnwashedMeme** [comment]
@alanjcastonguay   My apologies, we just have that disabled at the moment because it was causing too much trouble. I'll see if I can enable that at some point soon and watch it.

**Comment by 2opremio** [comment]
@UnwashedMeme have you seen this again recently?

**Comment by 2opremio** [comment]
Closing since I believe #1970 fixed the problem (released with Flux 1.12.1). Please reopen if it happens again.
